util/update_logdefs.ts

```diff
@@ -0,0 +1,374 @@
+  getFileList(): FileList {
+    const fileList: FileList = {
+      timelines: [],
+      triggers: [],
+    };
+
+    walkDirSync(path.posix.join(this.projectRoot, raidbossRelDir), (filepath) => {
+      if (/\/raidboss_manifest.txt/.test(filepath)) {
+        return;
+      }
+      if (/\/raidboss\/data\/.*\.txt/.test(filepath)) {
+        fileList.timelines.push(filepath);
+        return;
+      }
+      if (/\/raidboss\/data\/.*\.[jt]s/.test(filepath)) {
+        fileList.triggers.push(filepath);
+        return;
+      }
+    });
+
+    return fileList;
+  }
+
+  parseTriggerFile(file: string): void {
+    // We could dynamically import each trigger file and get the type from triggerSet,
+    // but we would lose the line number, which we need to pass the URL to the PR body,
+    // so just use regex to parse the file instead.
+
+    const contents = fs.readFileSync(file).toString();
+    const lines = contents.split(/\r*\n/);
+    const fileRegex = {
+      inTrSet: /^const triggerSet: TriggerSet<Data> = {/,
+      inTrArr: /^ {2}triggers: \[/,
+      inTrObj: /^ {4}\{/,
+      trType: /^ {6}type: '(?<type>[^']+)',$/,
+      outTrObj: /^ {4}\},/,
+      outTrArr: /^ {2}\],/,
+    };
+
+    let lineNum = 0;
+    let foundTriggerSet = false;
+    let foundTriggerArr = false;
+    let foundTrigger = false;
+    let insideTriggerObj = false;
+    let foundType = false;
+
+    for (const line of lines) {
+      ++lineNum;
+
+      if (line.match(fileRegex.inTrSet))
+        foundTriggerSet = true;
+      else if (foundTriggerSet && line.match(fileRegex.inTrArr)) {
+        foundTriggerArr = true;
+      } else if (foundTriggerArr && line.match(fileRegex.inTrObj)) {
+        insideTriggerObj = true;
+        foundTrigger = true;
+      } else if (insideTriggerObj && !foundType) {
+        const match = fileRegex.trType.exec(line);
+        const type = match?.groups?.type;
+        if (type !== undefined) {
+          foundType = true;
+          if (!this.isLogDefinitionName(type)) {
+            console.error(`ERROR: Missing log def for ${type} in ${file} (line: ${lineNum})`);
+            continue;
+          } else if (
+            this.logDefsNoInclude.includes(type) ||
+            this.logDefsNeverInclude.includes(type)
+          )
+            (this.triggerMatches[type] ??= []).push({
+              filename: file.replace(`${this.projectRoot}/`, ''),
+              line: lineNum,
+            });
+        }
+      } else if (foundType && line.match(fileRegex.outTrObj)) {
+        insideTriggerObj = false;
+        foundType = false;
+      } else if (foundTriggerArr && line.match(fileRegex.outTrArr))
+        break;
+    }
+
+    if (!foundTriggerSet || !foundTriggerArr || !foundTrigger)
+      console.error(`ERROR: Could not find triggers in ${file}`);
+  }
+
+  parseTimelineFile(file: string): void {
+    // TimelineParser doesn't have a convenient method for extracting an entry's type,
+    // so just process the file with regex.
```

Comment:
Could we not pull the line's decimal type from the constructed regex (via `^\^\(\?<type>([^)]+)\)`, remember to account for `Ability` being `2[12]`), and therefore leverage the already existing timeline parsing code?

We could then match this back to an actual def type trivially in log defs.

Reply:
Eesh.  It's a pretty creative solve, but I'm not wild about parsing regex with regex.  Also, I think your original concern -- about changes to the file format causing undetected breakage here -- would equally apply to regex'ing the timeline regex. But it did get me thinking about a potential alternative I want to run by you:

What if we add a no-op method to the TimelineParser class and call it at the right point during timeline parsing -- and then in this file, we do an extends class that overrides the method and extracts the type?  There would be negligible-to-no impact on runtime performance, I think, and a comment on the no-op method in the base class would help avoid possible breakage situations in the future.  Thoughts?

Reply:
> What if we add a no-op method to the TimelineParser class and call it at the right point during timeline parsing -- and then in this file, we do an extends class that overrides the method and extracts the type?

We do something similar for RaidEmulator at various points in the timeline and popup-text files, so I think this would be fine.